tidy3d/components/data/monitor_data.py

```diff
@@ -575,6 +578,150 @@ class FluxTimeData(MonitorData):
+
+    ux: ArrayLike[float, 2] = pd.Field(
+        ...,
+        title="Normalized kx",
+        description="Local x component of wave vectors for each diffraction order and frequency, "
+        "relative to ``local_origin`` and oriented with respect to ``u_axis``, "
+        "normalized by (2*pi/lambda) where lambda is the wavelength "
+        "associated with the background medium.",
+    )
+
+    uy: ArrayLike[float, 2] = pd.Field(
+        ...,
+        title="Normalized ky",
+        description="Local y component of wave vectors for each diffraction order and frequency, "
+        "relative to ``local_origin`` and oriented with respect to ``u_axis``, "
+        "normalized by (2*pi/lambda) where lambda is the wavelength "
+        "associated with the background medium.",
+    )
+
+    E: DiffractionDataArray = pd.Field(
+        ...,
+        title="E",
+        description="Complex components of the electric far field for each polarization "
+        "tangential to ``monitor.normal_axis``, in the local coordinate system whose "
+        " z-axis is ``monitor.normal_axis``.",
+    )
+    H: DiffractionDataArray = pd.Field(
+        ...,
+        title="H",
+        description="Complex components of the magnetic far field for each polarization "
+        "tangential to ``monitor.normal_axis``, in the local coordinate system whose "
+        "z-axis is ``monitor.normal_axis``.",
+    )
+
+    @property
+    def field_components(self) -> Dict[str, DataArray]:
+        """Maps the field components to thier associated data."""
+        # pylint:disable=no-member
+        return {field: getattr(self, field) for field in ["E", "H"]}
+
+    def normalize(self, source_spectrum_fn: Callable[[float], complex]) -> DiffractionData:
+        """Return copy of self after normalization is applied using source spectrum function."""
+        fields_norm = {}
+        for field_name, field_data in self.field_components.items():
+            src_amps = source_spectrum_fn(field_data.f)
+            fields_norm[field_name] = field_data / src_amps
+
+        return self.copy(update=fields_norm)
+
+    def _make_coords(self):
+        """Helper to make the coordinates dictionary."""
+        coords = {}
+        coords["orders_x"] = np.atleast_1d(self.E.orders_x.values)
+        coords["orders_y"] = np.atleast_1d(self.E.orders_y.values)
+        coords["polarization"] = np.array(["s", "p"])
+        coords["f"] = np.array(self.E.f.values)
+        return coords
+
+    @property
+    def angles(self):
+        """Return the (theta, phi) angles corresponding to each allowed pair of diffraction
+        orders as a data array. Disallowed angles are set to ``np.nan``.
+        """
+        # same dimensions as the fields except no polarization axis
+        thetas = np.zeros_like(self.E.sel(polarization="x").values, dtype=float)
+        phis = np.zeros_like(thetas, dtype=float)
+        for i, _ in enumerate(self.E.orders_x.values):
```

Comment:
would it help to vectorize this? or is It not advised?
I would think just using `kspace_2_sph` on `self.ux` and `self.uy` might work directly, even giving the nans.

Reply:
That can be done, but found that when numpy tries to compute the various trigonometric functions for arguments outside their domain, e.g., `arcsin(theta)` for `theta > 1`, it prints a bunch of error messages, even though the end result is fine (just filled with nans). So I just wanted to intercept any invalid cases and handle them manually. We could also try to suppress the numpy warnings / errors.

Another challenge was that `ux` and `uy` have different sizes and then there's the second frequency dimension, so I initially got errors in broadcasting. But maybe that's just a matter of adding `[None, ...]` extra dimensions myself before calling `kspace_2_sph`.

Reply:
on the broadcasting part, you may want to consider using `np.einsum('xf, yf -> xyf', ux**2, uy**2)` or yea sticking a `None` in `ux`, like `ux[:, None, :]**2 + uy[None, ...]**2`, maybe ensue is cleaner.

Reply:
on the warnings thing, I see what you mean when I try this locally.  Maybe you can try suppressing the warning? 

before:

```py
np.arcsin(10)
```
gives RuntimeWarning

but
```py
import warnings

with warnings.catch_warnings():
    warnings.filterwarnings('ignore')
    np.arcsin(10)
```

passes silently.